### PR TITLE
Add PiecewiseLinearNorm

### PR DESCRIPTION
**Pull request by dopplershift**
Renamed version of #3858.

Tests are failing on my machine due to my numpy 1.9.2 `interp` dropping masked arrays:

``` python
import numpy as np
vmin,vcenter,vmax = -1.0,1.5,4.0
x = [vmin, vcenter, vmax]
y = [0.0, 0.5, 1.0]
a = np.arange(-2, 5)
a = np.ma.array(a)
a[0] =np.ma.masked
np.interp(a, x, y)
```

I see:

```
array([ 0. ,  0. ,  0.2,  0.4,  0.6,  0.8,  1. ])
```

no mask

### CODE REVIEW

**Comment by jenshnielsen** [comment]
I have tested the numpy issue and it looks like it has been like this since at least 1.6.0

**Comment by efiring** [comment]
1) numpy interp: I think this is to be expected.  For the most part, plain numpy doesn't know how to handle masked arrays, which is why the ma module is so big.  Here, explicit handling seems to be in order--you need to do the interpolation using only the valid values.  The interp docstring says nothing about handling nans, either.
2) I think the idea for "piece-wise-linear" is that it is general, and can be applied with any number of segments.  The version here is a special case with two segments.  It would be nice to handle the general and special cases from the outset.
3) What about the inverse?  The basic linear Normalize norm is just a special case of piece-wise linear; the latter should be just as invertible as the former.

**Comment by tacaswell** [comment]
I am a bit nervous that the inverse was removed.

**Comment by OceanWolf** [comment]
Any update on this?

**Comment by OceanWolf** [comment]
I will take a look at this tomorrow and see what I can add.  Should I make a PR to merge into this branch, or work on top of this in a separate PR?

File: lib/matplotlib/colors.py [hunk 1]
```diff
@@ -963,6 +964,93 @@ def scaled(self):
...
+
+        vcenter : float, optional
+            The data value that defines ``0.5`` in the normalized data.
+            Defaults to halfway between *vmin* and *vmax*.
+
+        vmax : float, optional
+            The data value that defines ``1.0`` in the normalized data.
+            Defaults to the the max value of the dataset.
+
+        clip : bool, optional (default is False)
```
Comment (jkseppan) [inline lib/matplotlib/colors.py]: This parameter actually has no effect.

File: lib/matplotlib/colors.py [hunk 2]
```diff
@@ -963,6 +964,93 @@ def scaled(self):
...
+            raise ValueError("minvalue must be less than or equal to "
+                             "centervalue which must be less than or "
+                             "equal to maxvalue")
+        else:
+            vmin = float(vmin)
+            vcenter = float(vcenter)
+            vmax = float(vmax)
+            if clip:
+                mask = ma.getmask(result)
+                result = ma.array(np.clip(result.filled(vmax), vmin, vmax),
```
Comment (jkseppan) [inline lib/matplotlib/colors.py]: If clip is true, result is assigned here, but it's assigned again below.

File: lib/matplotlib/colors.py [hunk 3]
```diff
@@ -963,6 +964,93 @@ def scaled(self):
...
+                                  mask=mask)
+
+            x, y = [vmin, vcenter, vmax], [0, 0.5, 1]
+            # returns a scalar if shape == (1,)
```
Comment (jkseppan) [inline lib/matplotlib/colors.py]: Doesn't seem to hold at least for numpy 1.9.2:

```
In [29]: np.ma.masked_array(np.interp([5], [0,8,10], [0,0.5,1]))
Out[29]: 
masked_array(data = [ 0.3125],
             mask = False,
       fill_value = 1e+20)

In [30]: _.shape
Out[30]: (1,)
```

**Comment by jkseppan** [comment]
I added some suggestions on [my branch](https://github.com/jkseppan/matplotlib/tree/add-asym-norm), also submitted as a pull request to @dopplershift's branch.

**Comment by OceanWolf** [comment]
just looked at it dopplershift/matplotlib#2 but we still lack the generality... I think the API should read:
```python
def __init__(self, stops, clip=False):
```
We need stops as a kind of dictionary, not too sure on that yet.  Whether ``{<cmap-fraction>: <value>}`` or ``{<value>, <cmap-fraction>}``, I guess we would want an ordered dictionary, but as we still support ``python-2.6`` that means we should do it as a list of tuples... [(<cmap-fraction>, <value>), (<cmap-fraction>, <value>)], with this structure we have:
```python
class CenteredLinearNorm(PiecewiseLinearNorm):
  def __init__(self, vmin, vcenter, vmax, clip=False):
    stops = [(0., vmin), (0.5, vcenter), (1., vmax)]
    PiecewiseLinearNorm.__init__(self, stops, clip)
```

``PiecewiseLinearNorm`` will probably want to do some sort of validation to ensure order, in which case we can accept any kind of dict and build our own sorted dict.

**Comment by tacaswell** [comment]
I also started working on this a while ago, but got distracted.  I was envisioning an API something like

https://github.com/tacaswell/matplotlib/commit/de5eb530e0926a8ee34e8398623cd3bb87dfdfc8

**Comment by OceanWolf** [comment]
@tacaswell Any reason why you have them as two separate lists?  We have a 1:1 mapping, this would require an extra check that the two lists have equal lengths, and if we have two lists, we can easily put them into ordered dict format with ``zip(map_points, data_points)``.

**Comment by tacaswell** [comment]
Because they have different units and because there are less ways to input
this wrong.

If we take a list of tuples then we also need to take a 2D arrays, which
means we need to deal with the transpose, etc.

Also, this matches the underlying `np.intep` API so we can just pass
through to that and let numpy do those checks (we just need to check
monotonic as their docs explicitly say they do not check for that).

**Comment by OceanWolf** [comment]
Well I don't see different units as a problem... it underpins the whole definition of a dict {'time': 5}, string, int.

With "more ways to input it wrong" I think I see why you say that, because you think in terms of a list of tuples, but as I say we should think of it instead as passing in an ``OrderedDict``, or anything that ``stops = OrderedDict(stops)`` won't error, this approach feels better for the future when we finally drop support for python-2.6 (which you have already considered doing recently).

```python
def __init__(self, stops, clip=False):
  stops = OrderedDict(stops)  # will raise a ValueError on data
  stops = OrderedDict(sorted(stops.items(), key=lambda t: t[0]))  # sort keys
  map_points, data_points = stops.keys(), stops.values()  # get data ready for np.interp
  if not np.all(np.diff(data_points) > 0):
    raise ValueError("stops must increase monotonically")
```

We need far simpler error checking here because we utilize ordered dicts, this also makes the class a lot more useful as we simply duck-type the data input based on its dict-like structure.  Anything that can get cast to a dict, we accept.  While we don't support OrderedDict yet because of our support for python-2.6 we can for now cast to an ordinary dict and then extract and work internally with a list of tuples, this casting to a dict will mean we can also accept OrederedDicts right now, but internally handle them a bit more clumsily until we drop support for 2.6.

**Comment by jkseppan** [comment]
Are you sure there is a use case for a general piecewise linear mapping? Centering a diverging colormap's neutral color on a specific data value makes a lot of sense to me, but I'm not so sure about the more general case.

**Comment by OceanWolf** [comment]
@jkseppan I thought earlier about this earlier today, and I wondered about our accent colour map (a qualitative one), that might work well with data sources where we want to see the local variation, but the data contains big jumps in the data which would drown out that local variation otherwise...

**Comment by tacaswell** [comment]
@dopplershift was very excited about this for meteorology where they have color maps with multiple natural break points.  The other canonical use case would be to correctly map `gist_earth` on to elevation data.

**Comment by tacaswell** [comment]
@dopplershift @jkseppan This needs a rebase.  If this turns back into just a centered norm for now, that is cool it just needs to be renamed.

**Comment by jkseppan** [comment]
It seems that #3858 has a problem with finding a fitting name for the norm, and people liked PiecewiseLinearNorm, but that suggests more generality than implemented here. How about DivergingNorm, since it's likely to get used with a diverging colormap? Then we can later implement the more general case as PiecewiseLinearNorm if needed.

**Comment by OceanWolf** [comment]
I have just started on the generic version of this, we shall see how far I get...

**Comment by tacaswell** [comment]
Sold on diverging norm

**Comment by jklymak** [comment]
closing in lieu of #5054

**Pull request closed without merging**